File: frontend/src/api/audit.ts
import { api } from 'src/api/http';

// ── Types ───────────────────────────────────────────────────────────

export type AuditAction =
  | 'CREATE'
  | 'READ'
  | 'UPDATE'
  | 'DELETE'
  | 'ROLLBACK'
  | 'TRANSFER';

export interface AuditLogEntry {
  id: number;
  entity_type: string;
  entity_id: number;
  version: number;
  change_type: AuditAction;
  change_reason: string | null;
  changed_by: number | null;
  changed_by_display_name?: string | null;
  changed_at: string;
  handler_id: string;
  handled_ids: string[];
  ip_address: string;
  route_path: string | null;
  message_summary: string | null;
  sync_status: string | null;
  sync_error: string | null;
  synced_at: string | null;
}

export interface AuditLogDetail extends AuditLogEntry {
  data_snapshot: Record<string, unknown>;
  data_diff: {
    added: Record<string, unknown>;
    removed: Record<string, unknown>;
    changed: Record<string, { old: unknown; new: unknown }>;
  } | null;
}

export interface AuditStats {
  total_entries: number;
  creates: number;
  reads: number;
  updates: number;
  deletes: number;
}

export interface PaginationMeta {
  page: number;
  page_size: number;
  total: number;
}

export interface AuditLogListResponse {
  data: AuditLogEntry[];
  pagination: PaginationMeta;
}

export interface AuditQueryParams {
  user_id?: string;
  handler_id?: string;
  entity_type?: string;
  entity_id?: number;
  action?: AuditAction;
  date_from?: string;
  date_to?: string;
  search?: string;
  module?: string;
  page: number;
  page_size: number;
  sort_by: string;
  sort_desc: boolean;
}

// ── API functions ───────────────────────────────────────────────────

function toSearchParams(
  params: AuditQueryParams,
): Record<string, string | number | boolean> {
  return { ...params };
}

export async function fetchAuditLogs(
  params: AuditQueryParams,
): Promise<AuditLogListResponse> {
  return api
    .get('audit/activity', { searchParams: toSearchParams(params) })
    .json<AuditLogListResponse>();
}

export async function fetchAuditStats(
  params: AuditQueryParams,
): Promise<AuditStats> {
  return api
    .get('audit/stats', { searchParams: toSearchParams(params) })
    .json<AuditStats>();
}

export async function fetchAuditLogDetail(id: number): Promise<AuditLogDetail> {
  return api.get(`audit/activity/${id}`).json<AuditLogDetail>();
}

export async function exportAuditLogs(
  params: AuditQueryParams,
  format: 'csv' | 'json',
): Promise<Blob> {
  return api
    .get('audit/export', {
      searchParams: { ...toSearchParams(params), format },
    })
    .blob();
}

File: backend/app/schemas/audit.py
"""Audit log schemas for API request/response validation."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.models.audit import AuditChangeTypeEnum


class AuditLogEntry(BaseModel):
    """Single audit log entry returned by the API."""

    id: int
    entity_type: str
    entity_id: int
    version: int
    change_type: AuditChangeTypeEnum
    change_reason: str | None = None
    changed_by: int | None = None
    changed_by_display_name: str | None = None
    changed_at: datetime
    handler_id: str
    handled_ids: list[str] = Field(default_factory=list)
    ip_address: str
    route_path: str | None = None
    message_summary: str | None = None
    sync_status: str | None = None
    sync_error: str | None = None
    synced_at: datetime | None = None

    class Config:
        from_attributes = True


class AuditLogDetail(AuditLogEntry):
    """Full audit log entry with snapshot and diff (for detail view)."""

    data_snapshot: dict[str, Any] = Field(default_factory=dict)
    data_diff: dict[str, Any] | None = None

    class Config:
        from_attributes = True


class PaginationMeta(BaseModel):
    """Pagination metadata."""

    page: int
    page_size: int
    total: int


class AuditLogListResponse(BaseModel):
    """Paginated list of audit log entries."""

    data: list[AuditLogEntry]
    pagination: PaginationMeta


class AuditStats(BaseModel):
    """Summary statistics for audit logs."""

    total_entries: int = 0
    creates: int = 0
    reads: int = 0
    updates: int = 0
    deletes: int = 0
